### PR TITLE
Align WebSocket connection request with normal HTTP request at cvdr

### PR DESCRIPTION
**Pull request by k311093**
WebSocket connection needs to follow Proxy configuration and Authorization configuration to connect adb websocket backend through Cloud Orchestrator. Change to use same config and header as normal HTTP request does.

### CODE REVIEW

**Review by ikicha — APPROVED** [review]
Great! Thanks.

File: pkg/client/client.go [hunk 1]
```diff
@@ -86,6 +86,8 @@ type Service interface {
 	HostService(host string) HostOrchestratorService
 
 	RootURI() string
+
+	AuthHeader() (http.Header, error)
```
Comment (jemoreira) [inline pkg/client/client.go:90]: The code that establishes the websocket connection should be part of the `HostOrchestratorService`, so that it can be used from places other than the cli.

With that change there would be no need to expose the `AuthHeader` in the CO client, it can simply pass it to the HO client behind the scenes.

I understand this is a big change, so it's ok to do it in another PR.

**Review by 0405ysj — APPROVED** [review]
Thanks!

File: pkg/client/client.go (hunk 1 above)
Comment (k311093) [reply pkg/client/client.go]: Implemented WebSocket connection at the HO client and changed to use it when possible.

Thanks!

**Review by ser-io — CHANGES_REQUESTED** [review]
I think we can move the `PrepareAuth` refactoring into the last commit and have only one commit in this PR.

Having 3 commits were most of the changes in commit 1 and commit 2 are undo in commit 3 is misleading when looking at the history as well as getting the full picture of this PR.

**Comment by k311093** [comment]
> I think we can move the `PrepareAuth` refactoring into the last commit and have only one commit in this PR.
> 
> Having 3 commits were most of the changes in commit 1 and commit 2 are undo in commit 3 is misleading when looking at the history as well as getting the full picture of this PR.

Merged into one commit.

File: pkg/client/host_orchestrator_client.go [hunk 2]
```diff
@@ -114,6 +122,43 @@ func (c *HostOrchestratorServiceImpl) ConnectWebRTC(device string, observer wcli
...
+	}
+	if c.ProxyURL != "" {
+		proxyURL, err := url.Parse(c.ProxyURL)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse proxy %s: %w", c.ProxyURL, err)
+		}
+		dialer.Proxy = http.ProxyURL(proxyURL)
+	}
+
+	wsRoot := c.HTTPHelper.RootEndpoint
```
[the remark this answers is not in the document]
Comment (k311093) [reply pkg/client/host_orchestrator_client.go]: Changed to use URL object. Also used the object to join URL path.

File: pkg/client/httputils.go [hunk 3]
```diff
@@ -187,17 +187,25 @@ func (rb *HTTPRequestBuilder) JSONResDoWithRetries(ret any, retryOpts RetryOptio
 	return apiError
 }
 
-func (rb *HTTPRequestBuilder) doWithRetries(retryOpts RetryOptions) (*http.Response, error) {
+func (rb *HTTPRequestBuilder) PrepareAuth() error {
```
[the remark this answers is not in the document]
Comment (k311093) [reply pkg/client/httputils.go]: Changed function name into setAuthz.

**Comment by ser-io** [comment]
> > I think we can move the `PrepareAuth` refactoring into the last commit and have only one commit in this PR.
> > Having 3 commits were most of the changes in commit 1 and commit 2 are undo in commit 3 is misleading when looking at the history as well as getting the full picture of this PR.
> 
> Merged into one commit.

Good work. Thanks!

**Review by ser-io — APPROVED** [review]

File: pkg/cli/cli.go [hunk 4]
```diff
@@ -1319,26 +1319,36 @@ func runConnectionWebrtcAgentCommand(flags *ConnectFlags, c *command, args []str
...
+			InsecureSkipVerify: true,
+		},
 	}
-	b.WriteString(fmt.Sprintf("/devices/%s/adb", device))
-	return b.String()
+	wsConn, _, err := dialer.Dial(url.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to connect WebSocket %s: %w", url.String(), err)
+	}
+
```
Comment (ser-io) [inline pkg/cli/cli.go:1347]: nit: remove empty line.

**Review by ser-io — APPROVED** [review]

File: pkg/cli/cli.go (hunk 4 above)
Comment (k311093) [reply pkg/cli/cli.go]: Removed empty line.

Also removed other empty lines.
